## Description of the ** gen. nov

Etymology: 

Based on 16S rRNA gene sequence similarity, among its top 5 closest relatives,
there are representatives of the 
Desulfuromonadaceae, Geobacteraceae and Syntrophaceae families. 
with a high score of 86.09% (*Desulfuromonas carbonis*).
No POCP values >50% were identified with any of its closest relatives. 
Genome based assignment via GTDB-TK placed the genome under the 
WYAZ01 family of the Myxococcota phylum, supporting for a novel genus within the family.
<!-- As no representative of the Myxococcota phylum is present 
on the All-Species Living Tree project version currently used by Protologger. -->

<!-- Genome completeness is ;94.84
Genome contamination is ;3.89 -->

## Description of the ** sp. nov (MAG_197)

Etymology: 


KEGG analysis predicted 
acetate production from acetyl-CoA (EC:2.3.1.8, 2.7.2.1)
butyrate production from butanoyl-CoA (EC:2.8.3.8)
and 
propionate production from propanoyl-CoA (EC:2.3.1.8, 2.7.2.1). 
On top of that,
anoxygenic photosystem II was predicted from the genome (pufL, pufM),
biotin (vitamin B7) biosynthesis was predicted from pimeloyl-ACP/CoA (EC:2.3.1.47, 2.6.1.62, 6.3.3.3, 2.8.1.6)
and 
folate (vitamin B9) biosynthesis was predicted from 7,8-dihydrofolate (EC:1.5.1.3).
Riboflavin (vitamin B2) biosynthesis was predicted from GTP (EC:3.5.4.25, 3.5.4.26, 1.1.1.193, 3.1.3.104, 4.1.99.12, 2.5.1.78, 2.5.1.9, 2.7.1.26, 2.7.7.2) and
Cbb3-type cytochrome C oxidase was predicted based on the presence of sub-units I/II (ccoNO), III and IV. 
Within the genome, 21 flagellar proteins and 264 CAZymes were identified; 
starch was also predicted to be utilised as carbon source.
The G+C content of genomic DNA is 66.96%.
