## Description of the ** sp. nov (MAG_176)

Etymology:

Based on 16S rRNA gene sequence analysis its closest relatives were found to be members of the Gammaproteobacteria incertae sedis group with *Thiohalorhabdus denitrificans* having a best hit with a similiarity score of 91.72%. 
Genome based assignment via GTDB-TK placed the genome as a member of the Thiohalorhabdaceae family 
and 
ANI identified the input genome as novel, with the best match being to *Thiohalorhabdus denitrificans* with a value of 81.06%.
POCP analysis confirmed that the strain represents a novel species as it assigned it to *Thiohalorhabdus* with a POCP value of 68.02%.


KEGG analysis predicted siroheme biosynthesis from glutamate from the genome (EC:6.1.1.17, 1.2.1.70, 5.4.3.8, 4.2.1.24, 2.5.1.61, 4.2.1.75, 2.1.1.107/1.3.1.76/4.99.1.4)
Sulfide and L-serine were predicted to be utilised to produce L-cysteine and acetate (EC:2.3.1.30, 2.5.1.47)
Further, functional analysis predicted
L-glutamate production from ammonia via L-glutamine (EC:6.3.1.2, 1.4.1.-)
and
denitrification via conversion of nitrate into nitrogen (EC: 1.9.6.1, 1.7.2.1/1.7.99.1, 1.7.2.5, 1.7.2.4).
Biotin (vitamin B7) biosynthesis was predicted from pimeloyl-ACP/CoA (EC:2.3.1.47, 2.6.1.62, 6.3.3.3, 2.8.1.6)
and Cobalamin (vitamin B12) biosynthesis from cobinamide (EC:2.5.1.17, 6.3.5.10, 6.2.1.10, 2.7.1.156).
On top of that, folate (vitamin B9) biosynthesis was predicted from 7,8-dihydrofolate (EC:1.5.1.3)
and riboflavin (vitamin B2) biosynthesis from GTP (EC:3.5.4.25, 3.5.4.26, 1.1.1.193, 3.1.3.104, 4.1.99.12, 2.5.1.78, 2.5.1.9, 2.7.1.26, 2.7.7.2).
A total of 16 flagellar proteins were identified within the genome as well as 127 CAZymes; starch was predicted to be utilised as 
carbon source.
The G+C of the genomic DNA is 69.19%.

